connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkConnectPlanner.scala

```diff
@@ -2386,10 +2393,26 @@ class SparkConnectPlanner(val session: SparkSession) {
     }
 
     if (writeOp.hasForeachWriter) {
-      val foreach = writeOp.getForeachWriter.getPythonWriter
-      val pythonFcn = transformPythonFunction(foreach)
-      writer.foreachImplementation(
-        new PythonForeachWriter(pythonFcn, dataset.schema).asInstanceOf[ForeachWriter[Any]])
+      if (writeOp.getForeachWriter.hasPythonWriter) {
+        val foreach = writeOp.getForeachWriter.getPythonWriter
+        val pythonFcn = transformPythonFunction(foreach)
+        writer.foreachImplementation(
+          new PythonForeachWriter(pythonFcn, dataset.schema).asInstanceOf[ForeachWriter[Any]])
+      } else {
+        val foreachWriterPkt = unpackForeachWriter(writeOp.getForeachWriter.getScalaWriter)
+        val clientWriter = foreachWriterPkt.foreachWriter
+        if (foreachWriterPkt.rowEncoder == null) {
+          // rowEncoder is null means the client-side writer has type parameter Row,
+          // Since server-side dataset is always dataframe, here just use foreach directly.
+          writer.foreach(clientWriter.asInstanceOf[ForeachWriter[Row]])
+        } else {
+          val encoder = ExpressionEncoder(
+            foreachWriterPkt.rowEncoder.asInstanceOf[AgnosticEncoder[Any]])
```

Comment:
@hvanhovell is this safe? We are deserializing arbitrary Encoder class from user. Would it need user's jars? Otherwise, it could be a security issue.

Reply:
Should we instead always convert it to row encoder based on schema?

Reply:
If you could write the logic to work based on the schema of the dataset, pls do. The user's writer func would always expect a T, we need cast between row and T. It would be the best if we could not unpack the scala bytes (writer + encoder) before the execution, as the unpacking should be isolated.

However we know this issue with encoder and current impl and udfs. It is not necessary this PR's problem to address this issue.

Reply:
@zhenlineo agree. Should we try moving unpacking to execution time or doing this on server RPC thread is ok for now?

Reply:
It is okay for now.